node-wasm/src/node.rs

```diff
@@ -0,0 +1,254 @@
+use celestia_node::store::{IndexedDbStore, Store};
+use celestia_types::{hash::Hash, ExtendedHeader};
+use js_sys::Array;
+use libp2p::identity::Keypair;
+use libp2p::{identity, Multiaddr};
+use serde_wasm_bindgen::{from_value, to_value};
+use tracing::info;
+use wasm_bindgen::prelude::*;
+
+use crate::utils::js_value_from_display;
+use crate::utils::JsContext;
+use crate::utils::Network;
+use crate::wrapper::libp2p::NetworkInfo;
+use crate::Result;
+
+#[wasm_bindgen(js_name = Node)]
+struct WasmNode(Node<IndexedDbStore>);
+
+#[wasm_bindgen(js_name = NodeConfig)]
+pub struct WasmNodeConfig {
+    pub network: Network,
+    #[wasm_bindgen(skip)]
+    pub genesis_hash: Option<Hash>,
+    #[wasm_bindgen(skip)]
+    pub p2p_local_keypair: Keypair,
+    #[wasm_bindgen(skip)]
+    pub p2p_bootnodes: Vec<Multiaddr>,
+}
+
+#[wasm_bindgen(js_class = Node)]
+impl WasmNode {
+    #[wasm_bindgen(constructor)]
+    pub async fn new(config: WasmNodeConfig) -> Result<WasmNode> {
+        let network_id = network_id(config.network.into());
+        let store = IndexedDbStore::new(network_id)
+            .await
+            .js_context("Failed to open the store")?;
+
+        if let Ok(store_height) = store.head_height().await {
+            info!("Initialised store with head height: {store_height}");
+        } else {
+            info!("Initialized new empty store");
+        }
+
+        let node = Node::new(NodeConfig {
+            network_id: network_id.to_string(),
+            genesis_hash: config.genesis_hash,
+            p2p_local_keypair: config.p2p_local_keypair,
+            p2p_bootnodes: config.p2p_bootnodes,
+            p2p_listen_on: vec![],
+            store,
+        })
+        .await
+        .js_context("Failed to start the node")?;
+
+        Ok(Self(node))
+    }
+
+    /// Get node's local peer ID.
+    pub fn local_peer_id(&self) -> String {
+        self.0.local_peer_id().to_string()
+    }
+
+    /// Get current [`PeerTracker`] info.
+    pub fn peer_tracker_info(&self) -> Result<JsValue> {
+        let peer_tracker_info = self.0.peer_tracker_info();
+        Ok(to_value(&peer_tracker_info)?)
+    }
+
+    /// Wait until the node is connected to at least 1 peer.
+    pub async fn wait_connected(&self) -> Result<()> {
+        Ok(self.0.wait_connected().await?)
+    }
+
+    /// Wait until the node is connected to at least 1 trusted peer.
+    pub async fn wait_connected_trusted(&self) -> Result<()> {
+        Ok(self.0.wait_connected_trusted().await?)
+    }
+
+    /// Get current network info.
+    pub async fn network_info(&self) -> Result<NetworkInfo> {
+        Ok(self.0.network_info().await?.into())
+    }
+
+    /// Get all the multiaddresses on which the node listens.
+    pub async fn listeners(&self) -> Result<Array> {
+        let listeners = self.0.listeners().await?;
+
+        Ok(listeners
+            .iter()
+            .map(js_value_from_display)
+            .collect::<Array>())
```

Comment:
does it return an empty array or Err? We could as well just return here the or empty array directly as it's not possible to listen in browser afaik

Reply:
Currently this returns empty array `[]`, as expected. Is there a harm to leaving it as is? guess we could gain listening capabilities some day?

Reply:
we can leave it